### ISSUE TITLE
Non-negativity of initial condition during inversion

### ISSUE THREAD

**Issue by facusapienza21** [ODINN-SciML/ODINN.jl#375]
Following up on the conversation on #370 , I wanted to raise some math details regarding the constrain of non-negativity of the ice thickness. 

When optimizing the parameters of the component vector, we include both parameters of the neural network (unconstrained) and parameters of the initial condition (constrained). During inversion simulations, I observed the values of ice-thickness actually go negative, so it is important to address these. There are the few option I see as possible solutions. 

***(1) Change the parametrization***

We could instead of using $H_0$ as initial condition, use $\sigma(H_0)$ with $\sigma : R \mapsto R_{\geq 0}$. A natural choice will be $\sigma(x) = \log(1 + e^x)$ since this approximates the identity for positive values of $x$. 

*Implementation:* We just need to change how the initial condition is retrieved (instead of the entry of $\theta$, we apply first this function, which we can store in the initial condition) and change the adjoint calculation to account for the sensitivity $s_0 = \frac{\partial H_0}{\partial \theta_{H_0}} = \sigma' (\theta_{H_0})$.

***(2) Change the optimizier***

We could instead deal with this constraint directly in the optimizer (which I think is more natural, since the problem is an optimization with constraints, and the problem is not the gradient calculation, but rather the parameter update). This can be ugly, since it may required to rewrite part of the optimizer to account for constraints (e.g., making gradient projection steps), which may further lead to non-convergence as @albangossard mentioned. 

*Implementation:* This can be done with the Optim/Optimization API, or we can hardcode a version of this with callbacks in the optimizer and projection steps when the initial condition is called (a bit unclean, but it does the job and it is mathematically the same). This may not work for complex optimizers. 

***(3) Change the loss function***

Adding a barrier regularization term can do the job, by adding a regularization term that penalizes negative values (e.g., $max (0, -x)^2$). This will also do the job, but I am not a big fun of adding soft constraints when you can deal with this with a hard constrainted in the design and then avoid dealing with hyperparameter selection of the new regularization term and the "contamination" of this new term in the actual value of the loss function. 

*Implementation:* We can easely do this by adding a regularization term, which is in its ways of being developed and should be easy to integrate.
